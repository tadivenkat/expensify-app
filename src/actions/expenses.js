import database from '../firebase/firebase';

export const addExpenseAction = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        payload: expense
    };
};

export const startAddExpenseAction = (expenseData = {}) => {
    const {description = '', amount = 0, note = '', createdAt = 0} = expenseData;
    const expense = {description, amount, note, createdAt};
    return (dispatch) => {
        return database.ref('expenses').push(expense).then((data) => {
            dispatch(addExpenseAction({
                id: data.key,
                ...expense
            }));
        }).catch((error) => {
            console.log("Error adding expense", error);
        });
    }
};

export const removeExpenseAction = ({id} = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id 
    }
};

export const editExpenseAction = ({id, amount, description, note, createdAt} = {}) => {
    return {
        type: 'EDIT_EXPENSE',
        payload: {
            id,
            amount,
            description,
            note,
            createdAt
        }
    }
};

export const setExpensesAction = (expenses = []) => {
    return {
        type: 'SET_EXPENSES',
        payload: expenses
    }
};

export const startSetExpensesAction = () => {
    const expenses = [];
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpensesAction(expenses));
        }).catch((error) => {
            console.log("Error retreiving expenses", error);
        });
    };
};