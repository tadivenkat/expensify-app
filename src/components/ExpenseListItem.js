import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = (props) => {
    const {id, amount, createdAt, description, note} = props.expense;
    return (
        <div>
            <Link to={`/edit/${id}`}><h2>{description}</h2></Link>
            <p>Id: {id}</p>
            <p>Amount: {numeral(amount/100).format('$ 0,0.00')}</p>
            <p>CreatedAt: {!!createdAt ? moment(createdAt).format('DD-MMM-YYYY') : ''}</p>
            <p>Note: {note}</p>
        </div>
    );
}

export default ExpenseListItem;