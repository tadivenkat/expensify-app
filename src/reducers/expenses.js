
export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.payload];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.payload.id) {
                    return {...expense, ...action.payload};
                } else {
                    return expense;
                }
            });
        default: 
            return state;
    }
};