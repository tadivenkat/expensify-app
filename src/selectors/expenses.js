// Return subset of expenses applying the filters.

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        let startDateMatch = false;
        let endDateMatch = false;
        const textMatch = expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
        if (!!startDate) {
            if (!!expense.createdAt) {
                startDateMatch = startDate <= expense.createdAt;
            } else {
                startDateMatch = true;
            }
        } else {
            startDateMatch = true;
        }
        if (!!endDate) {
            if (!!expense.createdAt) {
                endDateMatch = endDate >= expense.createdAt;
            } else {
                endDateMatch = true;
            }
        } else {
            endDateMatch = true;
        }
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) => {
        switch (sortBy) {
            case 'amount':
                return expense1.amount - expense2.amount;
            case 'date':
                return expense1.createdAt - expense2.createdAt;
            default:
                return expense1.amount - expense2.amount;
        }
    });
};