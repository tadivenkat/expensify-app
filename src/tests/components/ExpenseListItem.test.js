import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('Testing ExpenseListItem Component Snapshot', () => {
    const wrapper = shallow(<ExpenseListItem expense={expenses[0]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});