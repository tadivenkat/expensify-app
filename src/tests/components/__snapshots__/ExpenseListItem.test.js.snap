// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ExpenseListItem Component Snapshot 1`] = `
<div
  className="card card-body mb-3"
>
  <Link
    to="/edit/1"
  >
    <h2>
      pen
    </h2>
  </Link>
  <p>
    Id: 
    1
  </p>
  <p>
    Amount: 
    $ 100.00
  </p>
  <p>
    CreatedAt: 
    01-Jan-1970
  </p>
  <p>
    Note: 
    test note for test expense
  </p>
</div>
`;
